### PR TITLE
2022 01 02 `hal_rrbot_control` port

### PR DESCRIPTION
**Pull request by zultron**
This PR is based on & follows #9, and is the last in a series of four to the first version working under ROS2.

This PR migrates the `hal_rrbot_control` package to ROS2, providing a complete, working demo of the `hal_hw_interface`.  Whereas before, the `hal_hw_interface` was derived from the `ros_control_boilerplate` repo, it is now built directly on `ros2_control`, and the `hal_rrbot_control` package now builds on the `ros2_control_demos` distro's packages, thus completing the migration off of the (ROS 1 only) `ros_control_boilerplate` distro and its (problematic) `rrbot_control` package.

The original demo included a gripper, now removed for simplicity.

The demo reuses the `rrbot_description` package `.xacro` files directly.  To do this, it requires patches parameterizing the `ros2_control` URDF `hardware` element, which specifies the hardware interface.  The package uses a temporary repo with these changes until the changes can be included upstream; some additional work is required to get those patches ready for a formal PR.

The demo also reuses as much of the rest of `ros2_control_demos` as is possible, paring this package down to a half-dozen files:  primarily the packaging, a launch file, and the HAL configuration.

### CODE REVIEW

**Comment by zultron** [comment]
This one is rebased on `foxy-devel`, checks passing, ready for review!

File: .github/workflows/upstream.rosinstall [hunk 1]
```diff
@@ -1,5 +1,13 @@
...
+    # FIXME The rrbot demo avoids duplicating the rrbot description
+    # package URDFs, which differ only in the ros2_control hardware
+    # tag, by passing the hardware plugin through parameters.  The
+    # patch parameterizing the rrbot demos is not yet in upstream
+    # packages.
+    #
+    # uri: https://github.com/ros-controls/ros2_control_demos.git
+    # version: foxy
+    uri: https://github.com/zultron/ros2_control_demos.git
+    version: foxy-dev-parameterize-hardware-plugin
```
Comment (vatanaksoytezer) [inline .github/workflows/upstream.rosinstall:13]: Why do we need ros2_control_demos? If it's only for rrbot, I don't mind having it here.

File: hal_rrbot_control/launch/rrbot.launch.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+            "hal_debug_output",
+            default_value="true",
+            description="Output HAL debug messages to screen (console).",
+        )
+    )
+    declared_arguments.append(
+        DeclareLaunchArgument(
+            "hal_debug_level",
+            default_value="1",
+            description="Set HAL debug level, 0-5.",
```
Comment (vatanaksoytezer) [inline hal_rrbot_control/launch/rrbot.launch.py:184]: Do you mind putting an explanation or linking a page here what different debug levels mean?

**Review by vatanaksoytezer — APPROVED** [review]
Just left two nit-picky comments. Lmk if those make sense, other than that lgtm. I'm going to approve this now and leave the merge to you as my comments can be addressed later, or not addressed at all.

File: hal_rrbot_control/launch/rrbot.launch.py (hunk 2 above)
Comment (zultron) [reply hal_rrbot_control/launch/rrbot.launch.py:184]: In practice, the default is 1, very little logging; when debugging HAL issues (e.g. HAL file problems, `hal_control_node` component problems), use 5, profuse debug messages.

File: .github/workflows/upstream.rosinstall (hunk 1 above)
Comment (zultron) [reply .github/workflows/upstream.rosinstall:13]: Yes, `ros2_control_demos` is a dependency of the `hal_rrbot_control` package in this repo.  That package isn't heavily tested other than by formatting checks and such, although down the road I'd ideally like to use it in an integration test just to check that the whole system can be brought up & play out a simple trajectory.

Comment (vatanaksoytezer) [reply .github/workflows/upstream.rosinstall:13]: 👍🏻 sgtm, feel free to merge whenever you are ready.

**Pull request merged**